sys/ztimer/core.c

```diff
@@ -117,6 +179,19 @@ uint32_t ztimer_set(ztimer_clock_t *clock, ztimer_t *timer, uint32_t val)
         clock->ops->set(clock, val);
     }
 
+    /* the clock is armed now
+     * everything down below doesn't impact timing */
+
+#if MODULE_ZTIMER_ONDEMAND
+    if (was_set) {
+        /* the given ztimer_t was set in the past
+         * remove the previously set instance */
+        _ztimer_release(clock);
+    }
+#else
+    (void)was_set;
+#endif
```

Comment:
If you take the suggestion to move and IRQ mangament to `_ztimer_release()`, the call could be moved below the `irq_restore()`. Briefly re-enabling IRQs in the middle would reduce the time with IRQs disabled and be beneficial for (soft) real time applications.

Reply:
Addressed by the fixups